### PR TITLE
Convert node side Generator

### LINKED ISSUE #18436: Add a ConstructSidesetsFromNodesetsGenerator
@hugary1995 s idea and it's pretty good imo

Some users add nodeset then want to rely on them to create sidesets. 
This is not possible right now because that operation is executed at the end of the mesh generation if `construct_sideset_from_nodeset = true`

@permcody I think we should keep track of meshing needs with a project or a label if we're going to get a student to add more capabilities in the future. What do you think?

### Discussed in https://github.com/idaholab/moose/discussions/18298

<div type='discussions-op-text'>

<sup>Originally posted by **BoZeng1997** July  9, 2021</sup>
Hi,

I have a strip with a meshed-in crack on the left in the centerline. I first defined two pieces of mesh for the top and bottom half and then stitched them together in the middle outside the crack. Then I want to redefine the two boundaries at the crack to have a closed sideset for later use. I have set `construct_side_list_from_node_list = true`, but there is no side sets generated from new boundaries. Please tell me what to do.

```
[Mesh]
  [top_half]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 94 #750/8
    ny = 16 #125
    xmax = 30 #30
    ymin = 0
    ymax = 5 #5
    boundary_id_offset = 0
    boundary_name_prefix = top_half
  []
  [top_nonstitch]
    type = BoundingBoxNodeSetGenerator
    input = top_half
    new_boundary = top_nonstitch
    bottom_left = '0 0 0' #5
    top_right = '5 0 0' #30
    show_info = true
  []
  [top_stitch]
    type = BoundingBoxNodeSetGenerator
    input = top_nonstitch
    new_boundary = top_stitch
    bottom_left = '5 0 0' #5
    top_right = '30 0 0' #30
    show_info = true
  []
  [bottom_half]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 94 #0
    ny = 16 #5
    xmax = 30 #0
    ymin = -5
    ymax = 0
    boundary_id_offset = 7
    boundary_name_prefix = bottom_half
  []
  [bottom_nonstitch]
    type = BoundingBoxNodeSetGenerator
    input = bottom_half
    new_boundary = bottom_nonstitch
    bottom_left = '0 0 0'
    top_right = '5 0 0'
    show_info = true
  []
  [bottom_stitch]
    type = BoundingBoxNodeSetGenerator
    input = bottom_nonstitch
    new_boundary = bottom_stitch
    bottom_left = '5 0 0'
    top_right = '30 0 0'
    show_info = true
  []
  [stitch]
    type = StitchedMeshGenerator
    inputs = 'top_stitch bottom_stitch'
    stitch_boundaries_pairs = 'top_stitch bottom_stitch'
    show_info = true
  []
  construct_side_list_from_node_list = true
[]
```
</div>

### PR DESCRIPTION
**Pull request by RocksSavage**
## Reason
#closes #18436

### CODE REVIEW

**Comment by GiudGiud** [comment]
Test failure is unrelated

any takers for reviewing @cticenhour @lindsayad @hugary1995

File: framework/src/meshgenerators/ConvertNodeSetSideSetGenerator.C [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  InputParameters params = MeshGenerator::validParams();
+
+  params.addClassDescription(
+      "Mesh generator which constructs side sets from node sets, and vise versa");
+  params.addRequiredParam<MeshGeneratorName>("input",
+                                             "Input mesh the operation will be applied to");
+  params.addRequiredParam<bool>("convert_node_list_from_side_list",
+                                "Flag signaling if nodes should be constructed from sides");
+  params.addRequiredParam<bool>("convert_side_list_from_node_list",
+                                "Flag signaling if sides should be constructed from nodes");
```
Comment (hugary1995) [inline framework/src/meshgenerators/ConvertNodeSetSideSetGenerator.C:31]: I am not a fan of using booleans to combine capabilities. The MOOSE-way of doing it is to have two separate classes. A `NodeSetToSideSetGenerator` and a `SideSetToNodeSetGenerator`. They could be deriving from the same parent class to reduce code duplication.

**Review by hugary1995 — CHANGES_REQUESTED** [review]
Some suggestions for starters.

**Comment by hugary1995** [comment]
Besides my comments above -- this is a really nice contribution!

File: framework/src/meshgenerators/ConvertNodeSetSideSetGenerator.C (hunk 1 above)
Comment (GiudGiud) [reply framework/src/meshgenerators/ConvertNodeSetSideSetGenerator.C:31]: On one hand selecting the nodesets / sidesets would be nice and make it more targeted.

On the other hand the other mesh generators that @RocksSavage are more important for our applications so I dont want to spend too long on these

So I'd say address the other comments and we'll advise later when we have good progress on the others

**Comment by cticenhour** [comment]
@GiudGiud I will give this another formal review when I return next week. @hugary1995 do you have any more comments?

**Comment by GiudGiud** [comment]
I ll clean it up today.
Please dont spend time reviewing before it s clean

**Comment by GiudGiud** [comment]
Ok let's see how this does with the test suite.

**Comment by hugary1995** [comment]
I'm being extra picky: I'd prefer the name `SidesetFromNodesetGenerator` over `NodesetToSidesetGenerator`. The current names probably give people a false impression that the source side set and node set are _removed_ as a consequence.

**Comment by GiudGiud** [comment]
but the docs say they are not :O

**Comment by hugary1995** [comment]
Fair enough

**Comment by GiudGiud** [comment]
The argument which could defeat me would be that all our generators are currently SideSetsFrom and not SideSetsTo

do you want to use this argument

**Comment by GiudGiud** [comment]
ok made the change.

Would love a thumbs up on this PR once it passes the distributed tests

**Review by cticenhour — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks for the reviews!

**Pull request merged**